ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Impl.hs

```diff
@@ -805,12 +817,29 @@ data IteratorHandle hash m = forall h. IteratorHandle
     -- exhausted and/or closed.
   , _it_end      :: !EpochSlot
     -- ^ The end of the iterator: the last 'EpochSlot' it should return.
-  , _it_end_hash :: !(Maybe hash)
+  , _it_end_hash :: !(IteratorEndHash hash)
     -- ^ The @hash@ of the last block the iterator should return. 'Nothing'
     -- when no @hash@ was specified, then only '_it_end' will be used to
     -- determine when to stop streaming.
   }
 
+data IteratorEndHash hash =
+    ItrNoEndHash
+  | ItrEndHash !hash
+  deriving NoUnexpectedThunks via UseIsNormalFormNamed "IteratorEndHash" (IteratorEndHash hash)
```

Comment:
Made this a separate type both so that we can override how we check this for thunks (we use `isNormalForm` so that we don't propagate anything up for `hash` -- the imm DB has enough constraints as it is), and also to make this thing actually strict in the hash.